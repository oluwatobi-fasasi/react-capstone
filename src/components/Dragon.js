// Dragon.js
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchDragonsAsync,
  cancelDragonReservation,
  reserveDragon,
  selectAllDragons,
} from '../redux/dragons/dragonsSlice';

export default function Dragon() {
  const dispatch = useDispatch();

  const dragons = useSelector(selectAllDragons);

  useEffect(() => {
    if (dragons.length === 0) {
      dispatch(fetchDragonsAsync());
    }
  }, [dispatch, dragons.length]);

  const reserveHandler = (id) => {
    dispatch(reserveDragon(id));
  };

  const cancelReservationHandler = (id) => {
    dispatch(cancelDragonReservation(id));
  };
  return (
    <div className="dragon-component">
      <ul>
        {dragons.map((dragon) => (
          <li key={dragon.id}>
            {dragon.flickr_images.length > 0 && (
              <img
                className="dragon-image"
                src={dragon.flickr_images}
                alt={`${dragon.name} Dragon`}
              />
            )}
            <div>
              <h2 className="dragon-name">{dragon.name}</h2>
              {dragon.reserved && <span className="reserved-badge-drag">Reserved</span>}
              <p className="dragon-desc">{dragon.desc}</p>
              <p>
                {dragon.reserved ? (
                  <button
                    type="button"
                    className="cancel-reservation  dragon-btn"
                    onClick={() => cancelReservationHandler(dragon.id)}
                  >
                    Cancel Reservation
                  </button>
                ) : (
                  <button
                    type="button"
                    className="reserve-rocket-btn dragon-btn"
                    data-testid="reserve-dragon-button"
                    onClick={() => reserveHandler(dragon.id)}
                  >
                    Reserve Dragon
                  </button>
                )}
              </p>
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
}
