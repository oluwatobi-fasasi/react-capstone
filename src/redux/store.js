import { configureStore } from '@reduxjs/toolkit';
import dragonSlice from './dragons/dragonsSlice';
import missionReducer from './mission/missionslice';
import rocketSlice from './rockets/rocketsSlice';

const store = configureStore({
  reducer: {
    dragon: dragonSlice,
    mission: missionReducer,
    rocket: rocketSlice,
  },
});

export default store;
