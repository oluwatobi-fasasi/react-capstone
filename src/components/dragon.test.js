import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/store';
import Dragon from './Dragon';

describe('Dragon Component', () => {
  it('renders Dragon name and description', () => {
    const main = renderer
      .create(
        <React.StrictMode>
          <Provider store={store}>
            <Router>
              <Dragon />
            </Router>
          </Provider>
        </React.StrictMode>,
      )
      .toJSON();

    expect(main).toMatchSnapshot();
  });
});
