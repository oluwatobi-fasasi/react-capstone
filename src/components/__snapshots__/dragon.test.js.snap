// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragon Component renders Dragon name and description 1`] = `
<div
  className="dragon-component"
>
  <ul />
</div>
`;
